### PR TITLE
feat: Add initial Password Input component

### PR DESCRIPTION
**Pull request by VincentSmedinga**
# Describe the pull request

Thank you for contributing to the project!
Please use this template to help us handle your PR smoothly.

## What

A component to enter a password.

This is a bare MVP version. It does not include a button to show or hide the characters – there is no design for it yet; even more so, the need for it hasn’t been discussed yet. I would say the features we have in our other inputs are a good basis for this one. People can start using this component, get the correct styling, and feedback any requirements they may have. I also created a ticket for the feature to show the password.

## Why

It was missing.

## How

- Scaffolded all files for a new component through plop
- Used `DateInput` and `TextInput` as inspiration
- Copied docs and tests from [an earlier approach](https://github.com/Amsterdam/design-system/commit/c5b75e04fded8b9b4caf2d3403c49daa2bc4e08d)

## Checklist

Before submitting your pull request, please ensure you have done the following. Check each checkmark if you have done so or if it wasn't necessary:

- [x] Add or update unit tests
- [x] Add or update documentation
- [x] Add or update stories
- [x] Add or update exports in index.\* files
- [x] Start the PR title with a Conventional Commit prefix, [as explained here](https://github.com/Amsterdam/design-system/blob/main/documentation/publishing.md?plain=1#L11).

## Additional notes

- Provide links to any related issues or discussions.
- Add a link to the specific story in the feature branch deploy.
- Mention any areas where additional review or feedback is needed.

### CODE REVIEW

File: packages/react/src/PasswordInput/PasswordInput.tsx [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+} & Omit<InputHTMLAttributes<HTMLInputElement>, 'aria-invalid' | 'type'>
+
+export const PasswordInput = forwardRef(
+  ({ className, invalid, ...restProps }: PasswordInputProps, ref: ForwardedRef<HTMLInputElement>) => (
+    <input
+      {...restProps}
+      aria-invalid={invalid || undefined}
+      className={clsx('ams-password-input', className)}
+      ref={ref}
+      type="password"
```
Comment (alimpens) [inline packages/react/src/PasswordInput/PasswordInput.tsx:22]: Should we set `autocapitalize="none"`, `autocorrect="off"` and `spellcheck="false"` here as well? It's a pretty firm guideline, no?

File: packages/css/src/components/password-input/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+# Password Input
+
+Helps users enter a password.
+
+Use this component when the input requires sensitive information, like passwords or PINs.
+It ensures that the input is not readable by others who might be looking at the screen.
+The characters entered are hidden, represented by squares.
+
+Consider setting the following attributes:
```
Comment (alimpens) [inline packages/css/src/components/password-input/README.md:11]: These are guidelines, no? Why the different format?

File: packages/css/src/components/password-input/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+1. Allow the user’s password manager to automatically fill the password through `autocomplete="current-password"`.
+   When asking for a new password, use `autocomplete="new-password"` instead.
+2. Add a `minlength` attribute to ensure passwords meet a minimum length requirement.
```
Comment (alimpens) [inline packages/css/src/components/password-input/README.md:15]:
```suggestion
2. Add a `minlength` attribute to ensure passwords meet a minimum length requirement.
  Describe this in the [Field](?path=/docs/components-forms-field--docs)’s description as well.
```

File: storybook/src/components/PasswordInput/PasswordInput.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      description: 'Prevents interaction. Avoid if possible.',
+    },
+  },
+} satisfies Meta<typeof PasswordInput>
+
+export default meta
+
+type Story = StoryObj<typeof meta>
+
+export const Default: Story = {}
```
Comment (alimpens) [inline storybook/src/components/PasswordInput/PasswordInput.stories.tsx:28]: Gov.uk advises against realtime validation. Maybe wrap the input in a `form` with `novalidate`?

File: packages/react/src/PasswordInput/PasswordInput.tsx [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  invalid?: boolean
+} & Omit<InputHTMLAttributes<HTMLInputElement>, 'aria-invalid' | 'type'>
+
+export const PasswordInput = forwardRef(
+  ({ className, invalid, ...restProps }: PasswordInputProps, ref: ForwardedRef<HTMLInputElement>) => (
+    <input
+      {...restProps}
+      aria-invalid={invalid || undefined}
+      className={clsx('ams-password-input', className)}
+      ref={ref}
```
Comment (alimpens) [inline packages/react/src/PasswordInput/PasswordInput.tsx:21]: Did you remove `dir={dir ?? 'auto'}` here on purpose? A password can be in a RTL language as well, no?

File: packages/react/src/PasswordInput/PasswordInput.test.tsx [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  it('supports ForwardRef in React', () => {
+    const ref = createRef<HTMLInputElement>()
+
+    const { container } = render(<PasswordInput ref={ref} />)
+
+    const component = container.querySelector(':only-child')
+
+    expect(ref.current).toBe(component)
+  })
+
```
Comment (alimpens) [inline packages/react/src/PasswordInput/PasswordInput.test.tsx:42]: Did you remove the following tests from TextInput on purpose? They seem relevant to me.

- should be working in a controlled state
- should not update the value when disabled

File: packages/react/src/PasswordInput/PasswordInput.test.tsx [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+          <PasswordInput id="password-field" />
+        </>,
+      )
+
+      const component = screen.getByLabelText(/password/i)
+
+      expect(component).toHaveAttribute('type', 'password')
+    })
+  })
+})
```
Comment (alimpens) [inline packages/react/src/PasswordInput/PasswordInput.test.tsx:84]: If we add `autocapitalize`, `autocorrect` and `spellcheck` by default, we should test for them as well.

File: packages/css/src/components/password-input/password-input.scss [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+  @include text-rendering;
+  @include reset;
+
+  &:hover {
+    box-shadow: var(--ams-password-input-hover-box-shadow);
+  }
+}
+
+.ams-password-input::placeholder {
+  color: var(--ams-text-input-placeholder-color);
```
Comment (alimpens) [inline packages/css/src/components/password-input/password-input.scss:40]:
```suggestion
  color: var(--ams-password-input-placeholder-color);
```

The token is missing as well.

File: packages/css/src/components/password-input/README.md (hunk 2 above)
Comment (VincentSmedinga) [reply packages/css/src/components/password-input/README.md:17]: Copied from Text Input, where it was an example. Adjusted.

File: packages/css/src/components/password-input/README.md (hunk 3 above)
Comment (VincentSmedinga) [reply packages/css/src/components/password-input/README.md:21]: Updated in a commit – we don’t (need to) use the `path` param.

File: packages/css/src/components/password-input/password-input.scss (hunk 8 above)
Comment (VincentSmedinga) [reply packages/css/src/components/password-input/password-input.scss]: Fixed and added in a commit.

File: packages/react/src/PasswordInput/PasswordInput.test.tsx (hunk 6 above)
Comment (VincentSmedinga) [reply packages/react/src/PasswordInput/PasswordInput.test.tsx:87]: No, not sure where I copied the tests from. Added now.

File: packages/react/src/PasswordInput/PasswordInput.tsx (hunk 5 above)
Comment (VincentSmedinga) [reply packages/react/src/PasswordInput/PasswordInput.tsx:24]: No not on purpose. It’s in now.

File: packages/react/src/PasswordInput/PasswordInput.tsx (hunk 1 above)
Comment (VincentSmedinga) [reply packages/react/src/PasswordInput/PasswordInput.tsx:26]: Sure thing!

File: storybook/src/components/PasswordInput/PasswordInput.stories.tsx (hunk 4 above)
Comment (VincentSmedinga) [reply storybook/src/components/PasswordInput/PasswordInput.stories.tsx:35]: Yes, added.

**Review by VincentSmedinga — COMMENTED** [review]
Thanks, thorough as always! 💪🏻

File: packages/react/src/PasswordInput/PasswordInput.tsx [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+import clsx from 'clsx'
+import { forwardRef } from 'react'
+import type { ForwardedRef, InputHTMLAttributes } from 'react'
+
+export type PasswordInputProps = {
+  /** Whether the value fails a validation rule. */
+  invalid?: boolean
+} & Omit<InputHTMLAttributes<HTMLInputElement>, 'aria-invalid' | 'type'>
```
Comment (alimpens) [inline packages/react/src/PasswordInput/PasswordInput.tsx:13]: We should probably omit `autocapitalize`, `autocorrect` and `spellcheck` here as well, right? So users don't get the impression they can change those?

File: storybook/src/components/TextInput/TextInput.docs.mdx [hunk 10]
```diff
@@ -26,7 +26,7 @@ On some devices, it may show an email-specific keyboard.
 Consider setting the following attributes:
 
 1. Set `autocomplete="email"` to help browsers autofill the user’s email address.
-2. Set `autocapitalize="none"`, `autocorrect="off"` and `spellcheck="false"` to stop browsers automatically changing user input.
+2. Set `autocapitalize="off"`, `autocorrect="off"` and `spellcheck="false"` to stop browsers automatically changing user input.
```
Comment (alimpens) [inline storybook/src/components/TextInput/TextInput.docs.mdx:29]: Apple recommends using `none`, they deprecated `off`

File: storybook/src/components/PasswordInput/PasswordInput.stories.tsx (hunk 4 above)
Comment (alimpens) [reply storybook/src/components/PasswordInput/PasswordInput.stories.tsx:35]: Hm, this doesn't seem to be working, it still validates while you type.

Comment (VincentSmedinga) [reply storybook/src/components/PasswordInput/PasswordInput.stories.tsx:35]: > The issue you’re encountering is likely due to the browser’s built-in CSS styling applied to the input field when the `minlength` attribute is used, regardless of the `novalidate` attribute on the form. The `novalidate` attribute disables the browser's built-in form validation, meaning it won’t prevent form submission or show default validation messages. However, it does not prevent the browser from applying its default styles based on the validation state of the input fields.
> 
> Here's what’s happening:
> 
> 1. The `minlength` attribute specifies a minimum number of characters that the user must enter. If the input value is shorter than this minimum, the input field is considered invalid.
> 2. Despite using `novalidate`, the browser still tracks the validity state of the input element and may apply default styles indicating an invalid state (such as turning the border red).

So I guess the issue here is our styling for `:invalid`, not the example itself.

File: storybook/src/components/TextInput/TextInput.docs.mdx (hunk 10 above)
Comment (VincentSmedinga) [reply storybook/src/components/TextInput/TextInput.docs.mdx]: Are you an encyclopaedia? (Serious question.) (Fixed.)

Comment (alimpens) [reply storybook/src/components/TextInput/TextInput.docs.mdx]: Ha! No, just crazy Google skills 😁

File: storybook/src/components/PasswordInput/PasswordInput.stories.tsx (hunk 4 above)
Comment (alimpens) [reply storybook/src/components/PasswordInput/PasswordInput.stories.tsx:35]: Hm, I see. Do we still want to recommend using it then? If it doesn't work the way we want it to? Both React Hook Form and Formio don't use the `minlength` attribute directly, I can imagine that's for this reason.

`minlength` is kind of similar to `maxlength` I guess, in that you can use them in form validation libraries, but not directly as attributes. Both validate too quickly, and in the case of `maxlength` it cuts off the input.

Comment (VincentSmedinga) [reply storybook/src/components/PasswordInput/PasswordInput.stories.tsx:35]: Yes, I think I agree. It still feels like a big leap to denounce browser validation entirely, but GOV.UK makes a compelling case and we all know how bad errors before finishing typing are.

I’ve removed the attribute and overturned the guideline.

Comment (alimpens) [reply storybook/src/components/PasswordInput/PasswordInput.stories.tsx:34]: Yeah, I think that is a good idea for now. We have to work this out in detail in our form guidelines, but that's something for later.

Without any validation attributes, we can get rid of the `<form noValidate>` decorator, I think?

**Review by alimpens — APPROVED** [review]

**Pull request merged**